### ISSUE TITLE
Ideas for a new ads client-side library

### ISSUE THREAD

**Issue by carlesandres** [Financial-Times/advertising#1]
This is a collection of the main attributes I believe a replacement for `o-ads` should have.

In my opinion, this approach is fully compatible with all the other ideas we've been discussing including:
- Sami's approach to refactoring the app
- Oliver's proposal of using a monorepo
- Alessandro and Andrew ideas for the future of ads-api 

# `ft-ads` main attributes
client-side library for ads rendering

### No interaction with targeting APIs
Should not interact with ads API, it should only provide a mechanism (callbacks or event listeners) for receiving and storing common targeting configuration that applies to all slots in the in-memory collection. And also another mechanism for receiving per-slot targeting config (and maybe storing it in memory).
 
### No declarative configuration
Should not allow any declarative initialisation. All the configuration and initialisation should be imperative, which will greatly reduce the complexity of the codebase, reduce the amount of code, remove entry barriers to new developers using the library and also for contributors to the project.

### Separate "dynamic" methods from "static" 
Should make a clear distinction between methods that required for fully server-side rendered pages (those whose ad slots are supposed to be initialised only once during the lifespan of the page) and dynamic pages (single-page apps, infinite scroll, etc...). The methods required for dynamic pages should be imported optionally so that they don't add unnecessary code to static pages.

### Closed source
Should be closed-source so that we don't have to be careful about FT-specific stuff code being leaked outside the company or caring about breaking functionality for users outside the FT group.

### No Origami
Should not be an origami component which will remove several constraints regarding tooling (e.g. origami-build-tools) and publishing. This will enable moving o-ads practices closer to what is used and expected from every other library in Customer Products. We can still publish via npm to anyone with access to @financial-times namespace. An example of problem we can solve is renaming `o-ads` to `@financial-times/o-ads` which will, in turn, fix npm-linking `o-ads` locally, which can be the cause of a lot of frustration since, for example, `n-ads` refers to the dependency as `@financial-times/o-ads`. Or `bower`

### Monorepo
Should be part of a monorepo that includes `n-ads` and `n-permutive` (I would even consider adding `ads-api`, and `n-live-ads-testing` to the same monorepo, I have experience with this architecture and it works really well, enabling greater productivity by having fewer code conflicts and fewer separate releases).

**Comment by oliverturner** [comment]
Re the monorepo, I think it might make sense to take inspiration from prior art (i.e. Page Kit) and create an `ad-kit` namespace to clearly indicate that these are related/interdependent packages:
```
@Financial-Times/ad-kit-ads
@Financial-Times/ad-kit-permutive
@Financial-Times/ad-kit-api
@Financial-Times/ad-kit-livetest
```

**Comment by carlesandres** [comment]
Good point @oliverturner ! I think we are supposed to keep using the `n-` prefix too as we are part of Customer Products (aka `next`).

**Comment by oliverturner** [comment]
This is where prior art is our friend: Page Kit demonstrates that this is not the case
https://github.com/Financial-Times/dotcom-page-kit/tree/master/packages

**Comment by carlesandres** [comment]
Fair enough. I believe the next wiki still advises to do so though.

**Comment by carlesandres** [comment]
Maybe the new client side library (`o-ads` replacement) should be called something like`ad-kit-client`. 
`n-ads` should make it obvious that it is for FT.com so somethinglike `ad-kit-dotcom` or `ad-kit-ftcom`.
And potentially the api should make it clear that it's meant for targeting like `ad-kit-targeting-api`

**Comment by oliverturner** [comment]
Yes, completely agree: one of the many nice things about this architecture is that it lets us create small modules named specifically for their use case 👍
